README.md
![monitorr](https://raw.githubusercontent.com/Monitorr/Monitorr/master/assets/img/monitorrbanner.png)


## - *Monitorr* is a webfront to live display the status of any webapp or service

[![GitHub Master-release](https://img.shields.io/github/release/monitorr/monitorr.svg?style=flat)](https://github.com/monitorr/monitorr/releases) [MASTER]

[![Docker build](https://img.shields.io/docker/build/monitorr/monitorr.svg?maxAge=2592000)](https://hub.docker.com/r/monitorr/monitorr/) [DOCKER]

[![GitHub (pre-)release](https://img.shields.io/github/release/monitorr/monitorr/all.svg)](https://github.com/monitorr/monitorr/releases) [DEVELOP]


**NOTICE (27 June 18)**: If updating to version **1.7** from ANY previous version, **BEFORE updating backup your custom.css file** located at: _`[Monitorr install path]/assets/data/css/custom.css`_ . After the update is complete, simply restore the custom.css file to the same location overwriting the custom.css file that was updated. This is due to a bug which has been resolved in version 1.7 therefore this process will no longer be needed in future updates. See the [WiKi Update Notice](https://github.com/Monitorr/Monitorr/wiki/NOTICE:-Updating-Monitorr) for more information.


## Features:

**Latest major change:** Multi HD display / Service ping response time display

- LIVE! (w/ option to pause live updating)
- Self-hosted.
- Responsive mobile display.
- Monitor any app/service on any domain (Curl as primary check, fsockopen as fallback).
- Enable/disable hot linking for each service.
- Service ping response time display. (NEW)
- Host system resources display (CPU, MEM, HD, PING, Uptime) (UPDATED)
- Integrated settings page w/ authentication.
- User customizable system threshold colors.
- Custom CSS tool. (NEW)
- Server DTG data. (UPDATED)
- Update tool via web UI / branch switching w/ update notification.
- Minimal UI for iFrame displays. (See: [WIKI](https://github.com/Monitorr/Monitorr/wiki/05-Integration:--Organizr))


**Features in development:**

- Python daemon (See: [Features 160](https://github.com/Monitorr/Monitorr/issues/160))
- Alerting (E-Mail, text, etc) (See: [Features 126](https://github.com/Monitorr/Monitorr/issues/126))


## Screenshots:

![Monitorr Screenshot](https://i.imgur.com/2wGLfLp.png)
<img src="https://i.imgur.com/UxDNcct.jpg?1" width="48%"> <img src="https://i.imgur.com/0smwRRu.jpg?1" width="50%">
<img src="https://i.imgur.com/BpBplql.png?1" width="32%">  <img src="https://i.imgur.com/PWWpDkn.jpg?1" width="32%"> <img src="https://i.imgur.com/Mnu0QJg.png?1" width="32%"> 


### Mobile:

![Monitorr Mobile SS](https://i.imgur.com/JLANkme.jpg?3)


## Prerequisites:

1) [PHP](https://secure.php.net/downloads.php) (7.1+ recommended)
2) [PHP cURL](https://secure.php.net/manual/en/book.curl.php)
3) [PHP ZipArchive](http://www.php.net/manual/en/zip.installation.php)
4) [PHP PDO](http://php.net/manual/en/book.pdo.php)
5) [SQLite](https://www.sqlite.org/index.html)
6) [GIT](https://git-scm.com/download/win) (Recommended for Windows hosts (see wiki)


## Quick Start:

- See full configuration instructions in the WiKi: https://github.com/Monitorr/Monitorr/wiki

1) Clone/download the Monitorr repository to your webserver. (Docker image available here: [![Docker build](https://img.shields.io/docker/build/monitorr/monitorr.svg?maxAge=2592000)](https://hub.docker.com/r/monitorr/monitorr/) )
2) Browse to: _`[localhost\domain]/monitorr/index.php`_
3) Establish data directory, and user database via secure registration tool.
4) Log in and configure Monitorr via secure settings UI.
5) Chill


## Feature Requests:

 [![Feature Requests](https://cloud.githubusercontent.com/assets/390379/10127973/045b3a96-6560-11e5-9b20-31a2032956b2.png)](https://feathub.com/Monitorr/Monitorr)

**Current feature requests:**

[![Feature Requests](https://feathub.com/Monitorr/Monitorr?format=svg)](https://feathub.com/Monitorr/Monitorr)


## Connect:

- Need live help?  Join us on Discord:   [![Discord](https://img.shields.io/discord/102860784329052160.svg)](https://discord.gg/YKbRXtt)

- E-mail: monitorrapp@gmail.com

- Buy us a beer! Donate:        [![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://paypal.me/monitorrapp)

- Check out our sister app **Logarr**:  https://github.com/Monitorr/Logarr


## About Us:

- [seanvree](https://github.com/seanvree) (Windows Wizard)
- [jonfinley](https://github.com/jonfinley) (Linux Dude)
- [wjbeckett](https://github.com/wjbeckett)


## Credits:

[ - Causefx](https://github.com/Causefx) - [christronyxyocum](https://github.com/christronyxyocum) - [rob1998](https://github.com/rob1998) - [Roxedux](https://github.com/si0972) -


## Libraries used in this project:

- [Alpaca](https://github.com/gitana/alpaca/)
- [PHP Login](https://github.com/panique/php-login-one-file)
- [ACE](https://github.com/ajaxorg/ace)
- [PACE](https://github.com/HubSpot/pace)
- [Form Validation](https://github.com/DrRoach/FormValidation)


changelog.md
<div id="about">

*   <div class="github-item">

    ## Develop: 1.7.1d

    <time class="github-item-time" datetime="2018-06-27T04:52:58Z"><span class="releasetime">Released on: 2018-06-27 at 04:52:58</span></time>

    ### - Summary:

    1\. This update is a MAJOR update to Monitorr with numerous fixes and enhancements.  
    2\. Docker image available here: [https://hub.docker.com/r/monitorr/monitorr/](https://hub.docker.com/r/monitorr/monitorr/)  

    ### - Update ALERT:

    If updating to version _1.7_* from ANY previous version, *_BEFORE updating backup your custom.css file_ located at: _(Monitorr install path)/assets/data/css/custom.css_ . After the update is complete, simply restore the custom.css file to the same location overwriting the custom.css file that was updated. This is due to a bug which has been resolved in version 1.7, therefore, this process will no longer be needed in future updates. See the [WiKi Update Notice](https://github.com/Monitorr/Monitorr/wiki/NOTICE:-Updating-Monitorr "WiKi Update Notice") for more information.  

    ### - Configuration changes:

    The following configuration changes need to be made AFTER updating to version 1.7.0 from ANY previous version. See the [WiKi Update Notice](https://github.com/Monitorr/Monitorr/wiki/NOTICE:-Updating-Monitorr "WiKi Update Notice") for more information.  

    a. _User Preferences:_  

    - Registration: Enable or disable access to the Registration tool. _NOTE: _This value should be changed to "Disable"_ AFTER updating._ [screenshot](https://user-images.githubusercontent.com/8906637/41943907-573ca978-795a-11e8-8f01-019c8b9eb9a4.PNG "screenshot")  

    b. _Monitorr Settings:_ [screenshot](https://user-images.githubusercontent.com/8906637/41945529-8df0cbd6-7962-11e8-87cf-2a92ca7e3374.PNG "screenshot")  

    - Time sync interval: Specifies how frequently (in milliseconds) the UI clock will synchronize time with the hosting webserver.  
    - HD display: Monitorr now has the much-anticipated feature of disabling or displaying multiple volumes (up to 3) in the system resources area of the UI.  
    - Ping color values: These values will determine the color of the ping indicators in the system resources are of the UI as well as individual service response times when enabled in "Services Configuration".  

    c. _Services Configuration:_  

    - Ping RT: Enables PING RT time output in the UI for each service. [screenshot](https://user-images.githubusercontent.com/8906637/41945612-dda1df8a-7962-11e8-8a59-82f31f0dc2b8.PNG "screenshot")  

    ### - CHANGE LOG:

    1. ADD: Ping response time for services. See: [#113](https://github.com/Monitorr/Monitorr/issues/113 "GitHub Issue")  
    2. ADD: User-defined HD volume / multiple volumes: see: [#148](https://github.com/Monitorr/Monitorr/issues/148 "GitHub Issue")  
    3. ADD: Monitorr release changelog. See: [#123](https://github.com/Monitorr/Monitorr/issues/123 "GitHub Issue")  
    4. FIX: Offline time does not coincide with the set timezone. See: [#165](https://github.com/Monitorr/Monitorr/issues/165 "GitHub Issue")  
    5. ADD: Disable browser output error if service is down. See: [#174](https://github.com/Monitorr/Monitorr/issues/174 "GitHub Issue")  
    6. CHANGE: New JS clock  
    7. ADD: execute single service check on settings.php load and when submit changes on Services Config settings page  
    8. CHANGE: Move UI icons from /img dir  
    9. FIX: Remove offline.json log files on service config settings page submit to prevent rouge offline.json files.  
    10. FIX: Prevent null data posting to json settings files  
    11. ADD: Offline service img and title will fade when offline.  
    12. CHANGE: correct scroll bars on settings pages  
    13. FIX: Uptime format on Windows  
    14. CHANGE: Change custom css script to local from CDN.  
    15. FIX: Update via UI will override custom CSS.  
    16. Add: Form validation to user input on _register form.  
    17. Add: CSS classes added to service images.  
    18. Change: date display on min site: to NOT show year.  
    19. Change: img upload error message / img upload modal color  
    20. Change: Clock will not sync when auto-update toggle is disabled.  
    21. Add: refresh marquee on settings page load.  
    22. ADD: MAIN PING indicator to respect color values.  
    </div>

    * * *

*   <div class="github-item">

    ## Develop: 1.6.4d

    <time class="github-item-time" datetime="2018-06-07T17:42:37Z"><span class="releasetime">Released on: 2018-06-07 at 17:42:37</span></time>

    This release is a _HOTFIX_ for a bug was discovered in the Monitorr settings page form.  

    ISSUE: If a user clicks inside a field on any of the settings pages and fires ENTER keystroke, the ajax call will write blank data to the user's .json files in their data directory thus completely breaking Monitorr if the UI is refreshed and the blank data is loaded into the DOM.  

    See [#167](https://github.com/Monitorr/Monitorr/issues/167 "GitHub Issue") for further details.

    </div>

    * * *

*   <div class="github-item">

    ## Master: 1.6.3m

    <time class="github-item-time" datetime="2018-06-07T17:16:40Z"><span class="releasetime">Released on: 2018-06-07 at 17:16:40</span></time>

    - This release is a HOTFIX for a bug was discovered in the Monitorr settings page form.  

    - ISSUE: If a user clicks inside a field on any of the settings pages and fires ENTER keystroke, the ajax call will write blank data to the user's .json files in their data directory thus completely breaking Monitorr if the UI is refreshed and the blank data is loaded into the DOM.  

    - See [#167](https://github.com/Monitorr/Monitorr/issues/167 "GitHub Issue") for further details.

    </div>

    * * *

*   <div class="github-item">

    ## Master: 1.6.1m


index.php
<!DOCTYPE html>
<html lang="en">

    <!--
    __  __             _ _
    |  \/  |           (_) |
    | \  / | ___  _ __  _| |_ ___  _ __ _ __
    | |\/| |/ _ \| '_ \| | __/ _ \| '__| '__|
    | |  | | (_) | | | | | || (_) | |  | |
    |_|  |_|\___/|_| |_|_|\__\___/|_|  |_|
            made for the community
    by @seanvree, @wjbeckett, and @jonfinley
    https://github.com/Monitorr/Monitorr
    -->

    <head>
        <meta charset="utf-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge">
        <link rel="manifest" href="webmanifest.json">
        <link rel="shortcut icon" type="image/x-icon" href="favicon.ico" />
        <link rel="apple-touch-icon" href="favicon.ico">

        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <meta name="description" content="Monitorr">
        <meta name="author" content="Monitorr">
        <meta name="version" content="php">
        <meta name="theme-color" content="#464646" />
        <meta name="theme_color" content="#464646" />

        <!-- Bootstrap core CSS -->
        <link href="assets/css/bootstrap.min.css" rel="stylesheet">

        <!-- Custom styles -->
        <link href="assets/css/main.css" rel="stylesheet">
        <link href="assets/data/css/custom.css" rel="stylesheet">

        <style>

            body {
                margin-bottom: 2vw;
                overflow-y: auto;
                overflow-x: hidden;
                background-color: #1F1F1F;
            }

            body::-webkit-scrollbar {
                width: .75rem;
                background-color: #252525;
            }

            body::-webkit-scrollbar-track {
                -webkit-box-shadow: inset 0 0 .25rem rgba(0, 0, 0, 0.3);
                box-shadow: inset 0 0 .25rem rgba(0, 0, 0, 0.3);
                border-radius: .75rem;
                background-color: #252525;
            }

            body::-webkit-scrollbar-thumb {
                border-radius: .75rem;
                -webkit-box-shadow: inset 0 0 .25rem rgba(0, 0, 0, .3);
                box-shadow: inset 0 0 .25rem rgba(0, 0, 0, .3);
                background-color: #8E8B8B;
            }

            body.offline #link-bar {
                display: none;
            }

            body.online #link-bar {
                display: block;
            }

            .auto-style1 {
                text-align: center;
            }

            #summary {
                margin-top: -2rem !important;
            }

            #header {
                margin-top: 2.5rem !important;
            }

            #hd {
                grid-column: 1/ span 2;
            }

            #services {
                margin-bottom: 7rem;
            }

            .pace .pace-activity {
                display: block;
                visibility: hidden;
                position: fixed;
                top: 40%;
                left: calc(50% - 13rem);
                width: 2rem;
                height: 2rem;
                margin-top: 1rem;
                border: solid .2rem transparent;
                border-top-color: #680233;
                border-left-color: #680233;
                border-radius: 1rem;
                -webkit-animation: pace-spinner 400ms linear infinite;
                -moz-animation: pace-spinner 400ms linear infinite;
                -ms-animation: pace-spinner 400ms linear infinite;
                -o-animation: pace-spinner 400ms linear infinite;
                animation: pace-spinner 400ms linear infinite;
                z-index: 3000;
            }

        </style>

        <script src="assets/js/jquery.min.js"></script>

            <!-- top loading bar function: -->
        <script src="assets/js/pace.js"></script>

         <?php

            $datafile = 'assets/data/datadir.json';
            $str = file_get_contents($datafile);
            $json = json_decode( $str, true);
            $datadir = $json['datadir'];
            $jsonfileuserdata = $datadir . 'user_preferences-data.json';

            if(!is_file($jsonfileuserdata)){

                $path = "assets/";

                include_once ('assets/config/monitorr-data-default.php');

                $title = $jsonusers['sitetitle'];

                $rftime = $jsonsite['rftime'];
            }

            else {

                $datafile = 'assets/data/datadir.json';

                include_once ('assets/config/monitorr-data.php');

                $title = $jsonusers['sitetitle'];
            }
         ?>

        <title>
            <?php
                echo $title . PHP_EOL;
            ?>
            | Monitorr
        </title>

            <!-- Clock functions: -->
        <script>
        
	        <?php
                //initial values for clock:
                $timezone = $jsonusers['timezone'];
                $dt = new DateTime("now", new DateTimeZone("$timezone"));
                $timeStandard = (int)($jsonusers['timestandard'] === "True" ? true : false);
                $timezone_suffix = '';
                if (!$timeStandard) {
                    $dateTime = new DateTime();
                    $dateTime->setTimeZone(new DateTimeZone($timezone));
                    $timezone_suffix = $dateTime->format('T');
                }
                $serverTime = $dt->format("D d M Y H:i:s");
	        ?>

            var nIntervId3;
            var onload;
            
            var serverTime = "<?php echo $serverTime;?>";
            var date = new Date(serverTime); 
            var timestandard = <?php echo $timeStandard;?>;
            var timeZone = "<?php echo $timezone_suffix;?>";
            var rftime = <?php echo $jsonsite['rftime'];?>;

            function updateTime() {
                setInterval(function () {
                    var res = date.toString().split(" ");
                    var time = res[4];
                    var timeSplit = time.split(":");
                    if(timestandard) {
                        time = parseInt((timeSplit[0] > 12) ? (timeSplit[0] - 12) : timeSplit[0]) + ":" + timeSplit[1] + ":" + timeSplit[2];
                        if(timeSplit[0] >= 12) {
                            time += " PM";
                        } else {
                            time += " AM";
                        }
                    }
                    var dateString = res[0] + ' | ' + res[2] + " " + res[1] + "<br>" + res[3];
                    var data = '<div class="dtg">' + time + ' ' + timeZone + '</div>';
                    data += '<div id="line">__________</div>';
                    data += '<div class="date">' + dateString + '</div>';
                    $("#timer").html(data);
                }, 1000);
            }

                // update UI clock with server time:

            function syncServerTime() {
                console.log('Monitorr time sync START | Interval: '+ rftime +' ms');
                $.ajax({
                    url: "assets/php/timestamp.php",
                    type: "GET",
                    timeout: 6000,
                    success: function (response) {
                        var response = $.parseJSON(response);
                        serverTime = response.serverTime;
                        timestandard = parseInt(response.timestandard);
                        timeZone = response.timezoneSuffix;
                        rftime = parseInt(response.rftime);
                        date = new Date(serverTime);
                        //setTimeout(function() {syncServerTime()}, rftime); //delay is rftime
                        $('#ajaxtimestamp').fadeOut();
                    },
                    error: function(x, t, m) {
                        if(t==="timeout") {
                            console.log("ERROR: Time sync timeout");
                            $('#ajaxtimestamp').fadeIn();
                        } else {
                            console.log("ERROR: timestamp failed");
                            $('#ajaxtimestamp').fadeIn();
                        }
                    }
                });
            }

            $(document).ready(function() {
                syncServerTime(); 
                updateTime();

                    //Stop clock update when refresh toggle is disabled:

                $(":checkbox").change(function () {

                    rftime =
                        <?php
                            $rftime = $jsonsite['rftime'];
                            echo $rftime;
                        ?>

                    if ($(this).is(':checked')) {
                        nIntervId3 = setInterval(syncServerTime, rftime); //delay is rftime
                    } 
                    else {
                        clearInterval(nIntervId3);
                    }
                });
            });

        </script>

            <!-- services status update function: -->
        <script type="text/javascript">

            var nIntervId;
            var onload;

            rfsysinfo =
                <?php
                    $rfsysinfo = $jsonsite['rfsysinfo'];
                    echo $rfsysinfo;
                ?>

            function statusCheck() {
                console.log('Service check START | Interval: <?php echo $rfsysinfo; ?> ms');
                $("#stats").load('assets/php/systembadges.php');
                $("#statusloop").load('assets/php/loop.php');
            };

                //Stop service status update when refresh toggle is disabled:

            $(document).ready(function () {
                $(":checkbox").change(function () {

                    if ($(this).is(':checked')) {
                        nIntervId = setInterval(statusCheck, rfsysinfo);
                    } 
                    else {
                        clearInterval(nIntervId);
                    }
                });
            });

        </script>

            <!-- Show loading modal pace indicator first onload only: -->
        <script>
            function showpace() {
                $('.pace-activity').addClass('showpace');
            }
        </script>

            <!-- marquee offline function: -->
        <script>

             var nIntervId2;
             var onload;

            $(document).ready(function() {

                $(":checkbox").change(function () {

                    var current = -1;
                    var onload;

                    function updateSummary() {

                        console.log('Service offline check START');

                        rfsysinfo =
                            <?php
                                $rfsysinfo = $jsonsite['rfsysinfo'];
                                echo $rfsysinfo;
                            ?>

                        $.ajax({
                            type: 'POST',
                            url: 'assets/php/marquee.php',
                            data: {
                                current: current
                            },
                            timeout: 7000,
                            success: function(data) {
                                if(data){
                                    result = $.parseJSON(data);
                                    console.log(result);
                                    $("#summary").fadeOut(function() {
                                        $(this).html(result[0]).fadeIn();
                                    });
                                    current = result[1];
                                }
                                else {
                                    current = -1;
                                    $("#summary").hide();
                                }
                                $('#ajaxmarquee').fadeOut();
                            },
                            error: function(x, t, m) {
                                if(t==="timeout") {
                                    console.log("ERROR: marquee timeout");
                                    $('#ajaxmarquee').fadeIn();
                                } else {
                                    console.log("ERROR: marquee failed");
                                    $('#ajaxmarquee').fadeIn();
                                }
                            }
                        });
                    }

                    if ($(this).is(':checked')) {
                        updateSummary();
                        nIntervId2 = setInterval(updateSummary, rfsysinfo);
                        console.log("Auto refresh: Enabled | Interval: <?php echo $rfsysinfo; ?> ms");
                    } else {
                        clearInterval(nIntervId2);
                        console.log("Auto refresh: Disabled");
                    }
                });
                $('#buttonStart :checkbox').attr('checked', 'checked').change();
            });

        </script>

            <!-- Load analog clock: -->
        <script src="assets/js/clock.js" async></script>

    </head>

    <body onload="statusCheck(), showpace()">

            <!-- Fade-in effect: -->
        <script>
            document.body.className += ' fade-out';
            $(function() {
                $('body').removeClass('fade-out');
            });
        </script>

            <!-- Append marquee alert if service is down: -->
        <div id="summary"></div>

            <!-- Ajax timeout indicator: -->
        <div id="ajaxtimeout">
            <div id="ajaxtimestamp" title="Time sync timeout. Refresh page." style="display: none;">
                <i class="fa fa-fw fa-exclamation-triangle"></i>
            </div>
            <div id="ajaxmarquee" title="Offline marquee timeout. Refresh page." style="display: none;">
                <i class="fa fa-fw fa-exclamation-triangle"></i>
            </div>
        </div>

        <div id="header">

            <div id="left" class="Column">
                <div id="clock">
                    <canvas id="canvas" width="120" height="120"></canvas>
                    <div id="timer"></div>
                </div>
            </div>

            <div id="center">

                <div id="centertext">
                    <a class="navbar-brand" href="
                        <?php
                            $siteurl = $jsonusers['siteurl'];
                            echo $siteurl . PHP_EOL;
                        ?>">
                        <?php
                            echo $title . PHP_EOL;
                        ?>
                    </a>
                </div>

                <div id="toggle">
                    <table id="slidertable">
                        <tr title="Toggle auto-refresh. Interval: <?php echo $rfsysinfo; ?> ms ">
                            <th id="textslider">
                            Auto Refresh:
                            </th>
                            <th id="slider">
                                <label class="switch" id="buttonStart">
                                    <input type="checkbox">
                                    <span class="slider round"></span>
                                </label>
                            </th>
                        </tr>
                    </table>
                </div>

            </div>

            <div id="right" class="Column">

                <div id="stats" class="container centered">
                    <!-- system badges go here -->
                </div>

            </div>

        </div>

            <!-- Check if datadir has been established: -->
        <?php

            $file = 'assets/data/datadir.json';

            if(!is_file($jsonfileuserdata)){

                echo '<div id="datdirerror">';
                    echo 'Data directory NOT detected. Proceed to <a href="settings.php" target="s" title="Monitorr Settings"> Monitorr Settings </a> and establish it.';
                echo '</div>';
            }

            else {
            }
        ?>

            <!-- Loading modal indicator: -->
        <div id="modalloadingindex" class="modalloadingindex" title="Monitorr is checking services.">
            
            <p class="modaltextloadingindex">Monitorr is loading ...</p>

        </div>

        <div id="services" class="container">

            <div class="row">
                <div id="statusloop">
                    <!-- loop data goes here -->
                </div>
            </div>

        </div>

        <div id="footer">

             <script src="assets/js/update_auto.js" async></script>

             <div id="settingslink">
                <a class="footer a" href="settings.php" target="s" title="Monitorr Settings"><i class="fa fa-fw fa-cog"></i>Monitorr Settings </a>
            </div>

            <p> <a class="footer a" href="https://github.com/monitorr/Monitorr" target="_blank" title="Monitorr Repo"> Monitorr </a> | <a class="footer a" href="https://github.com/Monitorr/Monitorr/releases" target="_blank" title="Monitorr Releases"> <?php echo file_get_contents( "assets/js/version/version.txt" );?> </a> </p>

            <div id="version_check_auto"></div>

        </div>

    </body>

</html>
